lib/tpm2_capability.c

```diff
@@ -33,38 +33,97 @@
 
 #include <tss2/tss2_sys.h>
 
+#include <string.h>
+
 #include "log.h"
 #include "tpm2_util.h"
 
 #include "tpm2_capability.h"
 
+static int tpm2_get_property_data_size(TPM2_CAP capability) {
+    int size;
+
+    switch (capability) {
+        case TPM2_CAP_ALGS:
+            size = sizeof(TPMS_ALG_PROPERTY);
+            break;
+        case TPM2_CAP_HANDLES:
+            size = sizeof(TPM2_HANDLE);
+            break;
+        case TPM2_CAP_COMMANDS:
+            size = sizeof(TPMA_CC);
+            break;
+        case TPM2_CAP_PP_COMMANDS:
+        case TPM2_CAP_AUDIT_COMMANDS:
+            size = sizeof(TPM2_CC);
+            break;
+        case TPM2_CAP_PCRS:
+            size = sizeof(TPMS_PCR_SELECTION);
+            break;
+        case TPM2_CAP_TPM_PROPERTIES:
+            size = sizeof(TPMS_TAGGED_PROPERTY);
+            break;
+        case TPM2_CAP_PCR_PROPERTIES:
+            size = sizeof(TPMS_TAGGED_PCR_SELECT);
+            break;
+        case TPM2_CAP_ECC_CURVES:
+            size = sizeof(TPM2_ECC_CURVE);
+            break;
+        case TPM2_CAP_VENDOR_PROPERTY:
+            size = sizeof(UINT32);
+            break;
+        default:
+            size = 0;
+            LOG_ERR("Unable to determine property size for capability:  %d\n", capability);
+            break;
+    }
+
+    return size;
+}
+
 bool tpm2_capability_get (TSS2_SYS_CONTEXT *sapi_ctx,
         TPM2_CAP capability,
         UINT32 property,
         UINT32 count,
         TPMS_CAPABILITY_DATA *capability_data) {
 
-    TPMI_YES_NO            more_data;
-
-    TSS2_RC rval = TSS2_RETRY_EXP(Tss2_Sys_GetCapability (sapi_ctx,
-                                NULL,
-                                capability,
-                                property,
-                                count,
-                                &more_data,
-                                capability_data,
-                                NULL));
-
-    if (rval != TPM2_RC_SUCCESS) {
-        LOG_ERR("Failed to GetCapability: capability: 0x%x, property: 0x%x",
-                 capability, property);
-        LOG_PERR(Tss2_Sys_GetCapability, rval);
-        return false;
-    } else if (more_data) {
-        LOG_WARN("More data to be queried: capability: 0x%x, property: "
-                 "0x%x\n", capability, property);
-        return false;
-    }
+    TPMI_YES_NO more_data;
+
+    TPMS_CAPABILITY_DATA local_data;
+
+    UINT32 offset = 0;
+    UINT32 property_size = tpm2_get_property_data_size(capability);
```

Comment:
See your coercing the int to UINT32 here, so probably just want to return UINT32.